### DISCUSSION TITLE
Custom security callback gets security bits param as 80 for RSA key sizes ranging from 920 to 1023

### DISCUSSION THREAD

**Discussion by vk18anubhav** [openssl/openssl#25283] [Q&A] [answered]
Hi,

We are setting a custom security callback in SSL_CTX_set_security_callback(). For TLS handshake, when certificate is having RSA key size of 920 to 1023, security bits param (int bits) in the security callback is coming as 80.

According to security levels, 80 bits security corresponds to key size 1024. Key size lesser than 1024 should have a value lesser than 80 for security bits param

**Comment by nhorman** [comment]
can you post the output of openssl x509 -in <cert> -text -noout for the certificate you are receiving?

**Comment by paulidale** [comment, marked as answer]
We determine the strength using the formula from NIST's [SP 800-56Br2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Br2.pdf) Appendix D: Maximum Security Strength Estimates for IFC Modulus Lengths.

> Let nBits = len(n), the bit length of the RSA modulus n included in a public key employed by the key-establishment scheme. The estimated maximum security strength E that can be supported by the modulus is determined using the following formula:
> $E = \frac{1.923 \sqrt[3]{nBits \cdot log_e(2)}\cdot(log_e(nBits \cdot log_e(2))^{2/3} - 4.69}{log_e(2)}$

There is some rounding involved since the result is not an integer, it always rounds favourably with respect to the strength (i.e. doesn't over estimate).

The code for this is here:
https://github.com/openssl/openssl/blob/32b43b9160cfcbb2940a0666869a680db827b892/crypto/rsa/rsa_lib.c#L305-L381

**Reply by vk18anubhav** [reply]
Hi Paul,

Thanks for your reply. I ran this program as standalone and here are results for key sizes ranging from 919 to 1024

bash-4.4$ ./strengthcheck 
key size = 919 , key strength = 72
key size = 920 , key strength = 80
key size = 1023 , key strength = 80
key size = 1024 , key strength = 80

Shouldn't the key sizes less than 1024 be having lesser strength than 80 even after round off ? if not, shall OpenSSL documentation be updated for strength 80 or security level 1 that the key sizes above 920 will be allowed ?

**Reply by mattcaswell** [reply]
So the documentation currently says this:

https://github.com/openssl/openssl/blob/8e7f39e8830ccafb41e52fbea895cb9740cebaec/doc/man3/SSL_CTX_set_security_level.pod?plain=1#L73-L78

Which is strictly speaking not accurate. It probably should say 920 rather than 1024. Although in practice you wouldn't normally encounter an RSA key size of 920. If its smaller than 1024 its usually 512. But the documentation could definitely be improved. Please feel free to submit a PR to update it if you would like.
